File: src/schema-to-xml.ts
import type { JsonSchema, ToolDefinition, XmlifyOptions } from "./types.js";

/**
 * Convert a JSON Schema tool definition into an XML schema description
 * that a model can understand. This produces a human/model-readable
 * XML template showing the expected structure.
 *
 * Example output:
 * ```xml
 * <tool name="browser" description="Browse a URL">
 *   <parameters>
 *     <url type="string" required="true">The URL to browse</url>
 *     <action type="string" enum="click,type,scroll">Action to perform</action>
 *   </parameters>
 * </tool>
 * ```
 */
export function toolToXmlSchema(
  tool: ToolDefinition,
  options: XmlifyOptions = {},
): string {
  const indent = options.indent ?? 2;
  const pad = (depth: number) => " ".repeat(depth * indent);

  const lines: string[] = [];
  lines.push(
    `<tool name="${escapeAttr(tool.name)}" description="${escapeAttr(tool.description)}">`,
  );
  lines.push(`${pad(1)}<parameters>`);

  if (tool.parameters.properties) {
    const required = new Set(tool.parameters.required ?? []);
    for (const [key, schema] of Object.entries(tool.parameters.properties)) {
      const paramXml = schemaPropertyToXml(
        key,
        schema,
        required.has(key),
        2,
        indent,
        options,
      );
      lines.push(paramXml);
    }
  }

  lines.push(`${pad(1)}</parameters>`);
  lines.push(`</tool>`);

  return lines.join("\n");
}

/**
 * Convert multiple tool definitions to XML.
 */
export function toolsToXmlSchema(
  tools: ToolDefinition[],
  options: XmlifyOptions = {},
): string {
  const parts = tools.map((t) => toolToXmlSchema(t, options));
  return `<tools>\n${parts.map((p) => indentBlock(p, options.indent ?? 2)).join("\n")}\n</tools>`;
}

function schemaPropertyToXml(
  name: string,
  schema: JsonSchema,
  isRequired: boolean,
  depth: number,
  indentSize: number,
  options: XmlifyOptions,
): string {
  const pad = " ".repeat(depth * indentSize);
  const attrs: string[] = [];

  if (options.typeHints && schema.type) {
    attrs.push(`type="${escapeAttr(schema.type)}"`);
  }
  if (isRequired) {
    attrs.push(`required="true"`);
  }
  if (schema.enum) {
    attrs.push(`enum="${escapeAttr(schema.enum.join(","))}"`);
  }
  if (schema.default !== undefined) {
    attrs.push(`default="${escapeAttr(String(schema.default))}"`);
  }

  const attrStr = attrs.length > 0 ? " " + attrs.join(" ") : "";

  // Object type — recurse into properties
  if (schema.type === "object" && schema.properties) {
    const lines: string[] = [];
    lines.push(`${pad}<${name}${attrStr}>`);
    const required = new Set(schema.required ?? []);
    for (const [key, childSchema] of Object.entries(schema.properties)) {
      lines.push(
        schemaPropertyToXml(
          key,
          childSchema,
          required.has(key),
          depth + 1,
          indentSize,
          options,
        ),
      );
    }
    lines.push(`${pad}</${name}>`);
    return lines.join("\n");
  }

  // Array type — show item template
  if (schema.type === "array" && schema.items) {
    const lines: string[] = [];
    lines.push(`${pad}<${name}${attrStr}>`);
    lines.push(
      schemaPropertyToXml(
        "item",
        schema.items,
        false,
        depth + 1,
        indentSize,
        options,
      ),
    );
    lines.push(`${pad}</${name}>`);
    return lines.join("\n");
  }

  // Leaf node — description as text content
  const desc = schema.description ? escapeXml(schema.description) : "";
  return `${pad}<${name}${attrStr}>${desc}</${name}>`;
}

function escapeAttr(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/"/g, "&quot;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

function escapeXml(s: string): string {
  return s
    .replace(/&/g, "&amp;")
    .replace(/</g, "&lt;")
    .replace(/>/g, "&gt;");
}

function indentBlock(block: string, size: number): string {
  const pad = " ".repeat(size);
  return block
    .split("\n")
    .map((line) => pad + line)
    .join("\n");
}

File: python/xmlify/schema_to_xml.py
"""Convert JSON Schema tool definitions to XML schema descriptions."""

from __future__ import annotations

from .types import JsonSchema, ToolDefinition, XmlifyOptions


def tool_to_xml_schema(tool: ToolDefinition, options: XmlifyOptions | None = None) -> str:
    """Convert a single tool definition to an XML schema description."""
    opts = options or XmlifyOptions()
    indent = opts.indent
    pad = lambda depth: " " * (depth * indent)

    lines: list[str] = []
    lines.append(f'<tool name="{_escape_attr(tool.name)}" description="{_escape_attr(tool.description)}">')
    lines.append(f"{pad(1)}<parameters>")

    props = tool.parameters.get("properties", {})
    required = set(tool.parameters.get("required", []))
    for key, schema in props.items():
        lines.append(_schema_property_to_xml(key, schema, key in required, 2, indent, opts))

    lines.append(f"{pad(1)}</parameters>")
    lines.append("</tool>")
    return "\n".join(lines)


def tools_to_xml_schema(tools: list[ToolDefinition], options: XmlifyOptions | None = None) -> str:
    """Convert multiple tool definitions to XML."""
    opts = options or XmlifyOptions()
    parts = [tool_to_xml_schema(t, opts) for t in tools]
    indented = "\n".join(_indent_block(p, opts.indent) for p in parts)
    return f"<tools>\n{indented}\n</tools>"


def _schema_property_to_xml(
    name: str,
    schema: JsonSchema,
    is_required: bool,
    depth: int,
    indent_size: int,
    options: XmlifyOptions,
) -> str:
    pad = " " * (depth * indent_size)
    attrs: list[str] = []

    if options.type_hints and schema.get("type"):
        attrs.append(f'type="{_escape_attr(schema["type"])}"')
    if is_required:
        attrs.append('required="true"')
    if schema.get("enum"):
        attrs.append(f'enum="{_escape_attr(",".join(schema["enum"]))}"')
    if schema.get("default") is not None:
        attrs.append(f'default="{_escape_attr(str(schema["default"]))}"')

    attr_str = (" " + " ".join(attrs)) if attrs else ""

    # Object type — recurse
    if schema.get("type") == "object" and schema.get("properties"):
        lines = [f"{pad}<{name}{attr_str}>"]
        req = set(schema.get("required", []))
        for key, child in schema["properties"].items():
            lines.append(_schema_property_to_xml(key, child, key in req, depth + 1, indent_size, options))
        lines.append(f"{pad}</{name}>")
        return "\n".join(lines)

    # Array type — show item template
    if schema.get("type") == "array" and schema.get("items"):
        lines = [f"{pad}<{name}{attr_str}>"]
        lines.append(_schema_property_to_xml("item", schema["items"], False, depth + 1, indent_size, options))
        lines.append(f"{pad}</{name}>")
        return "\n".join(lines)

    # Leaf node
    desc = _escape_xml(schema.get("description", ""))
    return f"{pad}<{name}{attr_str}>{desc}</{name}>"


def _escape_attr(s: str) -> str:
    return s.replace("&", "&amp;").replace('"', "&quot;").replace("<", "&lt;").replace(">", "&gt;")


def _escape_xml(s: str) -> str:
    return s.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")


def _indent_block(block: str, size: int) -> str:
    pad = " " * size
    return "\n".join(pad + line for line in block.split("\n"))
